### ISSUE TITLE
Combine functions (`xr.concat`, `xr.combine_by_coords`) create non-uniform chunk sizes

### ISSUE THREAD

**Issue by observingClouds** [pydata/xarray#9382]
### What happened?

I'm trying to concatenate two datasets with non-overlapping coordinates. Each coordinate in the dataset has the chunksize of 1. The merged dataset however has non-uniform chunks.

### What did you expect to happen?

Because the chunksizes of both datasets are uniform, even in case of the last chunk, I expect the chunks of the merged dataset to be uniform as well.

### Minimal Complete Verifiable Example

```Python
>>> import dask.array as da
>>> import xarray as xr

>>> data = da.zeros((5,10), chunks=(1,1))
>>> da1 = xr.DataArray(data=data, dims=['time', 'x'], coords={'time':range(5), 'x':range(10)})
>>> da2 = xr.DataArray(data=data, dims=['time', 'x'], coords={'time':range(5,10), 'x':range(10,20)})
>>> da1.chunks
((1, 1, 1, 1, 1), (1, 1, 1, 1, 1, 1, 1, 1, 1, 1))
>>> da2.chunks
((1, 1, 1, 1, 1), (1, 1, 1, 1, 1, 1, 1, 1, 1, 1))
>>> da1.time.values
array([0, 1, 2, 3, 4])
>>> da2.time.values
array([5, 6, 7, 8, 9])
>>> xr.concat([da1, da2], dim='time').chunks
((1, 1, 1, 1, 1, 1, 1, 1, 1, 1),
 (1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1))
>>> xr.combine_by_coords([da1,da2]).chunks
Frozen({'time': (1, 1, 1, 1, 2, 1, 1, 1, 1), 'x': (1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1)})
```

### MVCE confirmation

- [X] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [X] Complete example — the example is self-contained, including all data and the text of any traceback.
- [X] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [X] New issue — a search of GitHub Issues suggests this is not a duplicate.
- [X] Recent environment — the issue occurs with the latest version of xarray and its dependencies.

### Environment

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:45:18) [GCC 12.3.0]
python-bits: 64
OS: Linux
OS-release: 5.15.0-101-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.14.3
libnetcdf: 4.9.2

xarray: 2024.7.0
pandas: 2.2.2
numpy: 1.26.4
scipy: 1.14.0
netCDF4: 1.6.5
pydap: None
h5netcdf: None
h5py: 3.11.0
zarr: 2.18.2
cftime: 1.6.4
nc_time_axis: None
iris: None
bottleneck: None
dask: 2024.7.1
distributed: 2024.7.1
matplotlib: 3.9.1
cartopy: 0.23.0
seaborn: 0.13.2
numbagg: None
fsspec: 2024.6.1
cupy: None
pint: 0.24.3
sparse: None
flox: None
numpy_groupies: None
setuptools: 69.5.1
pip: 24.0
conda: 24.3.0
pytest: None
mypy: None
IPython: 8.26.0
sphinx: None

</details>

**Comment by observingClouds** [comment]
Thanks for your quick responses! As always great support!

The array is indeed filled with NaN where no data exists:
```
ds['zeros_like-6d7e7c771ea6fe3006a03a9616cc7e23'].values
Out[11]: 
array([[ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0., nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0., nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0., nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0., nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0., nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
       [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,  0.,  0.,  0., 0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,  0.,  0.,  0., 0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,  0.,  0.,  0., 0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,  0.,  0.,  0., 0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,  0.,  0.,  0., 0.,  0.,  0.,  0.,  0.,  0.,  0.]])
 ```

But more importantly, the issue is indeed fixed with updating from dask 2024.7.1 to 2024.8.1. Thanks @dcherian for the pointer. I should have checked it, but just recently created this environment. https://github.com/dask/dask/issues/5544 or https://github.com/dask/dask/pull/11269 for the win! :partying_face:

**Comment by dcherian** [comment]
Nice work @phofl !
